Mathlib/Order/Partition/Finpartition.lean

```diff
@@ -441,6 +441,15 @@ namespace Finpartition
 
 variable [DecidableEq α] {s t u : Finset α} (P : Finpartition s) {a : α}
 
+instance : Fintype (Finpartition s) where
+  elems := s.powerset.powerset.image
```

Comment:
This looks exponentially bad in terms of runtime / term size. Can we use the list function that was posted on Zulip which finds all partitions instead?

Reply:
Where?

Reply:
Edited above with a link

Reply:
The code in the Zulip link does not count the partitions I want. It merely takes in an ordered list and outputs all ways to put separators into it _preserving order_ – and the count of this is $2^{n-1}$ for a length-n list, not the Bell numbers.

Reply:
Here's the right function:
```
private def insertPerm (x : α) (l : List (List α)) : List (List (List α)) :=
  match l with
  | [] => [[[x]]]
  | y :: ys => ((x :: y) :: ys) :: (insertPerm x ys).map (y :: ·)

/-- Partitions of `l` into sublists. -/
def sublistPartitions (l : List α) : List (List (List α)) :=
  match l with
  | [] => [[]]
  | x :: xs => (sublistPartitions xs).bind fun p =>
    insertPerm x p
```

Reply:
I've opened #16647 with this definition; I am happy for you to try and solve the sorries there if you want to!

Reply:
I agree there's scope to improve the runtime of this definition, but let's get this current one merged while the details of the faster one are figured out? If I recall correctly, it's also difficult to show this function descends correctly to an operation on `Finset`s. Let's add a TODO in the file to your PR as well, so it's recorded.

Reply:
Agreed, a TODO is enough to unblock this PR.

Reply:
Done.